packages/ai-jsx/src/core/constrained-output.tsx

```diff
@@ -0,0 +1,141 @@
+/**
+ * This module defines affordances for constraining the output of the model
+ * into specific formats, such as JSON, YAML, or Markdown.
+ */
+import * as LLMx from '../index.js';
+import { ChatCompletion, SystemMessage, AssistantMessage, UserMessage } from './completion';
+import { memo } from './memoize';
+import yaml from 'js-yaml';
+
+function isJsonString(str: string): boolean | string {
+  try {
+    JSON.parse(str);
+  } catch (e: any) {
+    return e.message;
+  }
+  return true;
+}
+
+function isYamlString(str: string): boolean | string {
+  try {
+    // try to parse the string as YAML
+    yaml.load(str);
+  } catch (e: any) {
+    return e.message;
+  }
+  return true;
+}
+
+/**
+ * ChatCompletion that constrains the output to be a valid object format (e.g. JSON/YAML).
+ *
+ *
+ * @param type_name name of the type, to be used in the prompt
+ * @param validator function that returns true if the output is valid, or a string with an error message if not
+ *     // TODO: what should the signature of this function be?
+ * @param retries if output fails validation, how many times should we retry?
+ * @param verbose returns intermediate output if true
+ * @returns a string that validates as the given type (TODO: should I return the parsed object instead?)
+ */
+async function* ObjectFormatChatCompletion(
+  {
+    retries = 2,
+    verbose = false,
+    validator,
+    type_name,
+    children,
+    ...props
+  }: {
+    validator: (output: string) => boolean | string;
+    type_name: string;
+    retries?: number;
+    verbose?: boolean;
+    children: LLMx.Node[];
+  },
+  { render }: LLMx.RenderContext
+) {
+  const childrenWithCompletion = memo(
```

Comment:
I did that and I assumed it would work, but I noticed some weirdness at some points (I can't be sure, but I feel like the output would sometimes change; this happened very fast so I couldn't be sure). I didn't understand it, but I thought maybe that's why we have `memo` in the first place?

Reply:
Was it possible you had this instead?

```jsx
const children = <ChatCompletion>...</ChatCompletion>;
if (verbose) yield children
let output = await render(children)
```

That would cause independent renders because the thing being yielded is not via the render result but the tree itself.

Reply:
I feel like I was using the rendered version only, but I'm most likely wrong. It's good to know that I don't need `memo`. Removing it.